pkg/kubelet/kuberuntime/helpers.go

```diff
@@ -301,126 +302,166 @@ func getRuntimeServiceNameFromPodSpec(pod *v1.Pod) *string {
-	if remoteRuntimeEndpoints == "" {
-		return nil, nil, fmt.Errorf("runtimeEndpoints is empty")
+// Retrieve the runtime service with PODID
+func (m *kubeGenericRuntimeManager) GetRuntimeServiceByPodID(podId types.UID) (internalapi.RuntimeService, error) {
+	klog.V(4).Infof("Retrieve runtime service for podID %v", podId)
+	// firstly check the pod-runtimeService cache
+	if runtimeService, found := m.podRuntimeServiceMap[string(podId)]; found {
+		klog.V(4).Infof("Got runtime service [%v] for podID %v", runtimeService, podId)
+		return runtimeService, nil
 	}
 
-	runtimeServices := make(map[string]*runtimeService)
-	imageServices := make(map[string]*imageService)
+	// if not found in the cache, then query the runtime services
+	runtimeServices, err := m.runtimeRegistry.GetAllRuntimeServices()
+	if err != nil {
+		klog.Errorf("GetAllRuntimeServices failed: %v", err)
+		return nil, err
+	}
 
-	defer func() {
-		klog.V(4).Infof("runtime services: %v\n", runtimeServices)
-		for _, service := range runtimeServices {
-			klog.V(4).Infof("runtime service name:%s, workloadType:%s, endPointUrl:%s, isDefault:%v, serviceApi:%v",
-				service.name, service.workloadType, service.endpointUrl, service.isDefault, service.serviceApi)
+	var filter *runtimeapi.PodSandboxFilter
+
+	for _, runtimeService := range runtimeServices {
+		resp, err := runtimeService.ServiceApi.ListPodSandbox(filter)
+		if err != nil {
+			klog.Errorf("ListPodSandbox failed: %v", err)
+			return nil, err
 		}
-		klog.V(4).Infof("image services: %v\n", imageServices)
-		for _, service := range imageServices {
-			klog.V(4).Infof("image service name:%s, workloadType:%s, endPointUrl:%s, isDefault:%v, serviceApi:%v",
-				service.name, service.workloadType, service.endpointUrl, service.isDefault, service.serviceApi)
+
+		for _, item := range resp {
+			if item.Metadata.Uid == string(podId) {
+				klog.V(4).Infof("Got runtime service [%v] for podID %v", runtimeService, podId)
+				m.addPodRuntimeService(string(podId), runtimeService.ServiceApi)
+				return runtimeService.ServiceApi, nil
+			}
 		}
-	}()
+	}
 
-	endpoints := strings.Split(remoteRuntimeEndpoints, ";")
+	return nil, fmt.Errorf("failed find runtimeService with podId %v", podId)
+}
 
-	// Mostly legacy cluster agent
-	if len(endpoints) == 1 {
-		name, workloadType, endpointUrl, err := parseEndpoint(endpoints[0])
-		if err != nil {
-			return nil, nil, err
+// GetRuntimeServiceByPod returns the runtime service for a given pod from its SPEC
+// the GetRuntimeServiceByPod is called when POD is being created, i.e. the pod-runtime map and runtime service
+// will not have it
+func (m *kubeGenericRuntimeManager) GetRuntimeServiceByPod(pod *v1.Pod) (internalapi.RuntimeService, error) {
+	klog.V(4).Infof("Retrieve runtime service for POD %s", pod.Name)
+	runtimeName := getRuntimeServiceNameFromPodSpec(pod)
+
+	if runtimeName == nil || *runtimeName == "" {
```

Comment:
runtimeName is nil

https://github.com/futurewei-cloud/arktos/blob/99959f5ab4dffac8d8b852ca72d64eeedd6ff660/pkg/kubelet/kuberuntime/helpers.go#L301:6